src/csharp/Microsoft.Spark/Broadcast.cs

```diff
@@ -0,0 +1,160 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.IO;
+using System.Runtime.Serialization.Formatters.Binary;
+using Microsoft.Spark.Interop.Ipc;
+using Microsoft.Spark.Interop.Internal.Java.Util;
+using Microsoft.Spark.Interop;
+
+namespace Microsoft.Spark
+{
+    /// <summary>
+    /// A broadcast variable. Broadcast variables allow the programmer to keep a read-only variable
+    /// cached on each machine rather than shipping a copy of it with tasks. They can be used, for
+    /// example, to give every node a copy of a large input dataset in an efficient manner. Spark
+    /// also attempts to distribute broadcast variables using efficient broadcast algorithms to
+    /// reduce communication cost.
+    /// </summary>
+    [Serializable]
+    public sealed class Broadcast: IJvmObjectReferenceProvider
+    {
+        [NonSerialized]
+        private readonly JvmObjectReference _jvmObject;
+        private readonly long _bid;
+
+        internal Broadcast(SparkContext sc, object value, JvmObjectReference sparkContext)
+        {
+            var path = (string)Path.Combine(sc._temp_dir, Path.GetRandomFileName());
+            Version version = SparkEnvironment.SparkVersion;
+
+            // For Spark versions 2.3.0 and 2.3.1, Broadcast variable is created through different
+            // functions.
+            if (version.Major == 2 && version.Minor == 3 && (version.Build == 0 || 
+                version.Build == 1) )
+            {
+                WriteBroadcastValueToFile(sc, path, value);
+                var javaSparkContext = (JvmObjectReference)sparkContext.Jvm.CallStaticJavaMethod(
+                    "org.apache.spark.api.java.JavaSparkContext",
+                    "fromSparkContext",
+                    sparkContext);
+                _jvmObject = (JvmObjectReference)sparkContext.Jvm.CallStaticJavaMethod(
+                    "org.apache.spark.api.python.PythonRDD",
+                    "readBroadcastFromFile",
+                    javaSparkContext,
+                    path);
+            }
+            else
+            {
+                SparkConf sparkConf = sc.GetConf();
+                sc._encryption_enabled = bool.Parse(
+                    sparkConf.Get("spark.io.encryption.enabled",
+                    "false"));
+
+                var _python_broadcast = (JvmObjectReference)sparkContext.Jvm.CallStaticJavaMethod(
```

Comment:
What's up with this casing?

Reply:
Sorry slip of habit from python, should be camel case, will make the change. Thanks!